src/middlewared/middlewared/plugins/disk_/sync.py

```diff
@@ -208,18 +215,26 @@ async def sync_all(self, job):
                     # Do not issue unnecessary updates, they are slow on HA systems and cause severe boot delays
                     # when lots of drives are present
                     if self._disk_changed(disk, original_disk):
-                        await self.middleware.call('datastore.update', 'storage.disk', disk['disk_identifier'], disk)
-                        changed = True
+                        await self.middleware.call(
+                            'datastore.update', 'storage.disk', disk['disk_identifier'], disk, {'send_event': False}
+                        )
+                        changed.append({'id': disk['disk_identifier'], 'disk': disk})
                 else:
-                    await self.middleware.call('datastore.insert', 'storage.disk', disk)
-                    changed = True
+                    await self.middleware.call('datastore.insert', 'storage.disk', disk, {'send_event': False})
+                    changed.append({'id': disk['disk_identifier'], 'disk': disk})
 
         # make sure the database entries for enclosure slot information for each disk
         # matches with what is reported by the OS
         await self.middleware.call('enclosure.sync_disks')
 
-        if changed:
+        if changed or deleted:
             await self.middleware.call('disk.restart_services_after_sync')
+
+        for change in changed:
+            self.middleware.send_event('disk.query', 'CHANGED', id=change['id'], fields=change['disk'])
```

Comment:
Thank you, it's almost perfect now!

Unfortunately `change['disk']` now contains raw database fields: keys with `disk_` prefix, `disk.disk_extend` method not applied... API clients expect this to be the same as `disk.query` returns. I think the simplest way will be to re-query those objects using `call("disk.query", [["id", "in", [...]])`